### PR TITLE
tests: add codewhisperer integration tests

### PR DESCRIPTION
**Pull request by crossrac**
PR adding CodeWhisperer integration tests.

Adds integration test scenarios listed under 'Testing scenarios' in client side ops doc [here](https://quip-amazon.com/GeTtA7nHu9e1/Client-side-Integration-Tests#temp:C:KTA0c1ea8e817bc487094cac7c85).

Note:
- To run tests, use VSC launch config (select `Integration Tests` option and click `RUN AND DEBUG` - see screenshot below). Tests do not work with `Integration Tests (current file)` option.
- User must have a valid CodeWhisperer auth connection prior to running tests. If the user does not have a valid auth connection or if the tests are run using npm scripts, these integration tests will skip.

VSC launch config:
218815699-4b1c40bc-9eff-49fe-badd-73ed457cd223

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/integrationTest/codewhisperer/util.ts
Comment (justinmk3) [inline src/integrationTest/codewhisperer/util.ts:1]: Test util code should live in `src/integrationTest/util/`. That makes it clear the total size of the "test util" code, which should be minimal.

In fact, much of what's here already looks potentially generalizable for non-codewhisperer tests in the future.

**Review by justinmk3 — APPROVED** [review]

File: src/integrationTest/codewhisperer/referenceTracker.test.ts
Comment (leigaol) [inline src/integrationTest/codewhisperer/referenceTracker.test.ts:14]: If there is a model deployment, the response may not contain code reference because the recommendation will change. Is there any mitigation? If no, please be prepared.

Comment (andrewyuq) [reply src/integrationTest/codewhisperer/referenceTracker.test.ts:14]: Then we need to change, 1) the tests are not required for github approval flow 2) we can mark it as skipped until we posted a fix for manual runs

**Review by andrewyuq — APPROVED** [review]
Please make sure these run successfully before merge

**Comment by crossrac** [comment]
> Please make sure these run successfully before merge

Thanks all tests have been run successfully

File: src/integrationTest/codewhisperer/util.ts
Comment (crossrac) [reply src/integrationTest/codewhisperer/util.ts]: Thanks for your comment. That makes sense. I moved the `util` file to `src/integrationTest/util/`.

I also renamed `util` to `codewhispererUtil` to avoid confusion since it is only being used for codewhisperer auth connection now, but to your point this / the related imports can be easily updated if this logic is generalized or updated for other tests in the future.

File: src/integrationTest/codewhisperer/referenceTracker.test.ts
Comment (crossrac) [reply src/integrationTest/codewhisperer/referenceTracker.test.ts:22]: Thanks, I updated the `referenceTracker` tests to automatically skip for now and added the following comment to the top of the `referenceTracker.test.ts` file:

```
/*
New model deployment may impact references returned. 

These tests:
    1) are not required for github approval flow 
    2) will be auto-skipped until fix for manual runs is posted.
*/
```

File: src/integrationTest/codewhisperer/serviceInvocations.test.ts
Comment (JadenSimon) [inline src/integrationTest/codewhisperer/serviceInvocations.test.ts:39]: This should work without having to duplicate the condition across every test:
```suggestion
    beforeEach(function () {
        resetCodeWhispererGlobalVariables()
        RecommendationHandler.instance.clearRecommendations()
        if (!validConnection && this.currentTest) {
            this.currentTest.title += ` (skipped - no valid connection)`
            this.currentTest.skip()
        }
    })
```
You can make this into a util function too:
```ts
async function setConnectionOrSkipTest(ctx: Mocha.Context) {
    const conn = await setValidConnection()
    if (!conn && this.currentTest) {
       ctx.currentTest.title += ` (skipped - no valid connection)`
       ctx.currentTest.skip()
    }
}

// In `beforeEach`
validConnection = await setConnectionOrSkipTest(this)
```

File: src/integrationTest/codewhisperer/serviceInvocations.test.ts
Comment (JadenSimon) [inline src/integrationTest/codewhisperer/serviceInvocations.test.ts:122]: No need to mock a text editor. Do this instead:
```ts
const doc = await vscode.workspace.openTextDocument(vscode.Uri.file(filename))
const editor = await vscode.window.showTextDocument(doc)
```

**Review by JadenSimon — APPROVED** [review]
> Tests do not work with Integration Tests (current file) option.

It _should_ work, but you need to have the test file open/focused in the editor.

File: src/integrationTest/codewhisperer/serviceInvocations.test.ts
Comment (crossrac) [reply src/integrationTest/codewhisperer/serviceInvocations.test.ts]: Thanks for this input. I added this logic to the util file.

I changed the function slightly so that the valid connection is still set in a different function, because I found that if I make the `beforeEach` async and include the `skip` there I get the error `sync skip; aborting execution`.

so the updated function is:

```
export function skiptTestIfNoValidConn(validConnection: boolean, ctx: Mocha.Context) {
    if (!validConnection && ctx.currentTest) {
        ctx.currentTest.title += ` (skipped - no valid connection)`
        ctx.currentTest.skip()
    }
}
```
invocation: 

```
before(async function () {
        validConnection = await setValidConnection()
    })

beforeEach(function () {
        ...
        skiptTestIfNoValidConn(validConnection, this)
    })
```

**Review by andrewyuq — APPROVED** [review]

**Comment by justinmk3** [comment]
Can all of these e2e-related files live in a dedicate folder such as:

    src/integrationTest/codewhisperer/e2e

Later, we will relocated all e2e tests to a dedicated top-level folder (like `src/teste2e/`). Meanwhile, grouping all e2e tests in one place helps that later relocation.

**Comment by crossrac** [comment]
> Can all of these e2e-related files live in a dedicate folder such as:
> 
> ```
> src/integrationTest/codewhisperer/e2e
> ```
> 
> Later, we will relocated all e2e tests to a dedicated top-level folder (like `src/teste2e/`). Meanwhile, grouping all e2e tests in one place helps that later relocation.

Hi @justinmk3 , no problem. I moved the test files to `src/integrationTest/codewhisperer/e2e`

**Comment by crossrac** [comment]
> > Please make sure these run successfully before merge
> 
> Thanks all tests have been run successfully

Follow up to confirm tests have been run successfully after additional subsequent changes.

**Review by justinmk3 — APPROVED** [review]

**Pull request merged**
